src/db/spanner/models.rs

```diff
@@ -783,6 +796,152 @@ impl SpannerDb {
+        } else {
+            Ok(results::GetQuotaUsage::default())
+        }
+    }
+
+    pub async fn touch_user_collections(
+        &self,
+        user: &HawkIdentifier,
+        collection_id: i32,
+    ) -> Result<()> {
+        // This will also update the counts in user_collections, since `update_collection_sync`
+        // is called very early and only updates the timestamp.
+        // This will also write the tombstone if there are no records and we're explicitly
+        // specifying a TOMBSTONE collection_id.
+        // This function should be called after any write operation.
+        let mut total_bytes = 0;
+        let mut count = 0;
+        let timestamp = self.timestamp()?;
+        let mut set_sql = "";
+
+        self.metrics
+            .clone()
+            .start_timer("storage.quota.update_existing_totals", None);
+        let calc_sql = if self.quota_enabled {
+            "SELECT SUM(BYTE_LENGTH(payload)), COUNT(*)
+            FROM bsos
+           WHERE fxa_uid = @fxa_uid
+             AND fxa_kid = @fxa_kid
+             AND collection_id = @collection_id
+           GROUP BY fxa_uid"
+        } else {
+            "SELECT COUNT(*)
+            FROM bsos
+           WHERE fxa_uid = @fxa_uid
+             AND fxa_kid = @fxa_kid
+             AND collection_id = @collection_id
+           GROUP BY fxa_uid"
+        };
+        let result = self
+            .sql(calc_sql)?
+            .params(params! {
+                "fxa_uid" => user.fxa_uid.clone(),
+                "fxa_kid" => user.fxa_kid.clone(),
+                "collection_id" => collection_id.to_string(),
+            })
+            .execute_async(&self.conn)?
+            .one_or_none()
+            .await?;
+        if let Some(result) = result {
+            // Update the user_collections table to reflect current numbers.
+            if self.quota_enabled {
+                total_bytes = result[0]
+                    .get_string_value()
+                    .parse::<i64>()
+                    .map_err(|e| DbErrorKind::Integrity(e.to_string()))?;
+                count = result[1]
+                    .get_string_value()
+                    .parse::<i32>()
+                    .map_err(|e| DbErrorKind::Integrity(e.to_string()))?;
+                set_sql = "UPDATE user_collections
+                SET modified = @modified,
+                    count = @count,
+                    total_bytes = @total_bytes
+                WHERE fxa_uid = @fxa_uid
+                    AND fxa_kid = @fxa_kid
+                    AND collection_id = @collection_id";
+            } else {
+                set_sql = "UPDATE user_collections
+                SET modified = @modified
+                WHERE fxa_uid = @fxa_uid
+                AND fxa_kid = @fxa_kid
+                AND collection_id = @collection_id";
+            };
+        } else {
+            // we're explicitly setting a tombstone.
```

Comment:
Couple edge cases here that I think are problematic:

- Given a user_collection that already exists but does not have any bsos, the SELECT would return no results (None), triggering an INSERT into user_collections. Which would error out w/ ALREADY_EXISTS
- Given a user_collection that doesn't already exist, it seems like this no longer inserts (touches) user_collections unless TOMBSTONE was specified

If that's the case I'm surprised no tests caught this -- might be worth adding some that would?

Reply:
Ok, refactored this function a fair bit. as part of it:
1) it no longer looks for explicit tombstones, since tombstones are just another collection type and are handled as such.
2) will ALWAYS have a `set_sql` so will ALWAYS perform some action.

I proved that tests are hitting all cases both with and without `SYNC_ENABLE_QUOTA`.

Reply:
Do we need any additional tests for these cases?

Reply:
We shouldn't no, because the existing tests are already exercising the paths and checking the results.